CJAUnit 5 - Firebase Realtime Database Implementation

frankiebarrios opened the pull request:
## What It Does
This pull request is set to complete mini-project: add a database of unit 5 of the ux-onboarding-project.

This work is taking place to complete mini-project: add a database of unit 5 of the ux-onboarding-project.

**Implementation:**
- Installed, configured, and implemented Firebase Realtime Database.
- Rewrote LocalStorage API to utilize Firebase Database.

## How To Test
- Install dependencies using instruction in README.
- Start server using yarn start or build using yarn build.

## Notes/Caveats

None
## Checklist

Under penalty of public shaming, I avow that I:
**Many items in list N/A** 
- [X] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [X] Verified that there are no compiler or linter errors or warnings
- [X] Verified the build in production(optimized) mode
- [X] Updated the docs, if necessary
- [X] Included the appropriate labels
- [X] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
**Many items in list N/A** 
- [X] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

## Dependencies
1. Install Node and npm
2. Install Yarn
3. Install Ux-Lint

Commit:
- Installed Firebase SDK.
- Configured Firebase Realtime Database.

Commit:
Clean Up

frankiebarrios committed:
Update README.md

Commit:
- Reading from Firebase Database successfully within <intern-app> component

Commit:
Revert "- Reading from Firebase Database successfully within <intern-app> component"

This reverts commit 98807a72726cd829bb7ebee25be3fa1484ecbb39.

Commit:
Merge branch 'Unit-5' of github.com:frankielb/frankie-barrios-ux-iop into Unit-5

Commit:
- Moved Firebase Data Load from within component to API.

frankiebarrios committed:
Update localStorage.js

Commit:
Checkpoint

Commit:
- Beginning of API rewrite using Firebase Database vs Local Storage.

Commit:
Merge branch 'Unit-5' of github.com:frankielb/frankie-barrios-ux-iop into Unit-5

# Conflicts:
#	src/api/localStorage.js

Commit:
Disconnected localStorage API and connected Firebase API to App.
Completed successful write with new API from component.

Commit:
- Created firebase versions of deleteAllUsers(), getAllUsers(), and createUser(). Testing successfully.
- Corrected issue with user data not pulling in locally with getAllUsers().
- Refactored createUser() to correct issue with data structure in database upon user creation.
- Made some UI changes to accommodate new API changes.
- Clean up.

Commit:
- Created  updateUser(), getUserIndex(), generateId(), syncData(), and getUser() for Firebase API,  testing successfully.
- Added user.id back to use for backend processing. Keeping prop hidden from UI.

Commit:
- Created deleteUser().
- Refactored syncData() helper to work with getData() and updateUser().
- Changed how deleteUser() is called.
- Clean up.

frankiebarrios committed:
Clean up

Commit:
Checkpoint:
- Rough form validation working. Will need refactored.

Commit:
UI changes

kddrake commented on src/api/firebaseAPI.js:
```diff
@@ -0,0 +1,94 @@
+      .on('child_added', function (snapshot) {
+        match = snapshot.key
+      });
+    firebase
```
what happens here if a match isn't found?

kddrake commented on src/api/firebaseAPI.js:
```diff
@@ -0,0 +1,94 @@
+      const firebaseData = data.val();
+      const keys = Object.keys(firebaseData);
+      const userObj = [];
+      for (let i = 0; i < keys.length; i++) {
```
this for could be simplified
```suggestion
      keys.forEach(key => { userObj.push(firebaseData[key]) });
```

kddrake commented on src/api/firebaseAPI.js:
```diff
@@ -0,0 +1,94 @@
+      }
+      this.users = userObj;
+    } catch (error) {
+      console.log('Error Pulling In Users: ', error);
```
if you're going to log errors to the console use `console.error()`

kddrake commented on src/components/intern-app.html:
```diff
@@ -3,6 +3,10 @@
+<!-- The core Firebase JS SDK is always required and must be listed first -->
+<script src="https://www.gstatic.com/firebasejs/6.4.0/firebase-app.js"></script>
+<!-- Firebase Database Link -->
+<script src="https://www.gstatic.com/firebasejs/5.0.1/firebase-database.js"></script>
```
you should try adding these your your `package.json` over in-line scripting. I assume it's possible, though I don't fully know for sure. I think Christa was able to.

frankiebarrios replied:
Turned out I didn't need that at all. Removed it.

Commit:
- Added error catching to deleteUser().
- Removed unused Firebase <scripts> from <intern-app>.
- Changed error logging method.
- Refactored getData().

kddrake commented on src/components/intern-app.html:
```diff
@@ -51,25 +65,21 @@
-        this.users = this.storage.getAllUsers();
+
+      updateUserList() {
+        this.users = this.storage.getAllUsers();      
```
get rid of this trailing whitespace

kddrake commented on src/components/create-user.html:
```diff
@@ -115,18 +124,79 @@ <h1 class="div-center">Create User</h1>
-      updateUserList() {
-        this.createUser();
+      deleteAllUsers() {
+        this.storage.deleteAllUsers();        
```
lets get rid of this trailing whitespace here, too

kddrake reviewed (changes requested):
The validation work in here should really be in it's own PR (or in a friction clean up PR) as it doesn't really have anything to do with your database.

Commit:
- Removed validation work. Belongs to another branch.

Commit:
Removed trailing whitespace.

kddrake reviewed (changes requested):
run through it in the browser and check your console, seeing a couple of errors. Edit isn't working, either.

Commit:
- Resolved issues causing app not to function.

kddrake commented on src/components/intern-app.html:
```diff
@@ -51,25 +65,21 @@
-        this.users = this.storage.getAllUsers();
+
+      updateUserList() {
+        this.users = this.storage.getAllUsers();   
```
need to remove this trailing whitespace

kddrake reviewed (changes requested):
your create-user component looks like it has friction work in it. needs to be removed, or you need to convince me it needs to stay in.

Commit:
no message

Commit:
- Removing UI Friction work from branch.

kddrake commented on src/components/create-user.html:
```diff
@@ -74,22 +72,25 @@
-          <input type="text" placeholder="Your email.." value="{{user.email::input}}">
+          <label>Phone</label>
+          <input type="text" id="phone" placeholder="Your phone number.." value="{{user.phone::input}}">
+          
```
still got some whitespace here that'll need removed.

kddrake reviewed (changes requested):
Should probably pull the list of users on init. Kinda weird to have to click a button.

frankielb commented:
> Should probably pull the list of users on init. Kinda weird to have to click a button.

@kddrake Have made that change in the UI Friction branch.

Commit:
- Removed trailing whitespace.
